flocker/testtools/_flaky.py

```diff
@@ -22,10 +43,274 @@ def flaky(jira_key):
+                if result_type == _ResultType.skip:
+                    result.addSkip(case, details=details)
+                    skip_reported = True
+
+            if not skip_reported:
+                self._output.write(
+                    '@flaky(%s): passed %d out of %d runs '
+                    '(min passes: %d; max runs: %d)'
+                    % (case.id(), successes, len(results), flaky.min_passes,
+                       flaky.max_runs)
+                )
+
+                result.addSuccess(case, details=combined_details)
+        else:
+            # XXX: How are we going to report on tests that sometimes fail,
+            # sometimes error, sometimes skip? Currently we just error.
+            result.addError(case, details=combined_details)
+        result.stopTest(case)
+        return result
+
+    def _attempt_test(self, case):
+        """
+        Run 'case' with a temporary result.
+
+        :param testtools.TestCase case: The test to run.
+
+        :return: a tuple of ``(successful, result, details)``, where
+            ``successful`` is a boolean indicating whether the test was
+            succcessful, ``result`` is a _ResultType indicating what the test
+            result was and ``details`` is a dictionary of testtools details.
+        """
+        tmp_result = testtools.TestResult()
+        self._run_test(case, tmp_result)
+        result_type = _get_result_type(tmp_result)
+        details = pmap(case.getDetails())
+        if result_type == _ResultType.skip:
+            # XXX: Work around a testtools bug where it reports stack traces
+            # for skips that aren't passed through its supported
+            # SkipException: https://bugs.launchpad.net/testtools/+bug/1518100
+            [reason] = list(tmp_result.skip_reasons.keys())
+            details = details.discard('traceback').set(
+                'reason', text_content(reason))
+        _reset_case(case)
+        return (tmp_result.wasSuccessful(), result_type, details)
+
+
+class _ResultType(Names):
+    """
+    Different kinds of test results.
+    """
+
+    success = NamedConstant()
+    error = NamedConstant()
+    failure = NamedConstant()
+    skip = NamedConstant()
+    unexpected_success = NamedConstant()
+    expected_failure = NamedConstant()
+
+
+def _get_result_type(result):
+    """
+    Get the _ResultType for ``result``.
+
+    :param testtools.TestResult result: A TestResult that has had exactly
+        one test run on it.
+    :return: A _ResultType for that result.
+    """
+    if result.testsRun != 1:
+        raise ValueError('%r has run %d tests, 1 expected' % (
+            result, result.testsRun))
+
+    if len(result.errors) > 0:
+        return _ResultType.error
+    elif len(result.failures) > 0:
+        return _ResultType.failure
+    elif len(result.expectedFailures) > 0:
+        return _ResultType.expected_failure
+    elif len(result.unexpectedSuccesses) > 0:
```

Comment:
:grey_question: I've never understood what expectedFailures and unexpectedSuccesses are for... but should unexpectedSuccesses be a failure condition and thus higher priority than expectedFailures? This is at most a nit, and more probably a casual musing.